### PR TITLE
Migrate TestAddonMoreInfo from Enzyme

### LINKED ISSUE #11158: Migrate /components/TestAddonMoreInfo.js to react-testing-library
Migrate /components/TestAddonMoreInfo.js to react-testing-library

### PR DESCRIPTION
**Pull request by bobsilverberg**
Fixes #11158 and includes some setup for using react-testing-library.

### CODE REVIEW

File: tests/unit/helpers.js [hunk 1]
```diff
@@ -1470,3 +1474,68 @@ export const createExperimentData = ({
...
+  if (!elements.length) {
+    return null;
+  }
+  if (elements.length > 1) {
+    throw new Error(`More than one element found for tag ${tagName}`);
+  }
+  return elements[0];
+};
+
+export const render = (ui, options = {}) => {
```
Comment (bobsilverberg) [inline tests/unit/helpers.js:1508]: Some notes:

- Many (in fact most) of our components use the `translate` HOC which ends up injecting `i18n` into the component. This helper function wraps the component under test with `I18nProvider` which does the same thing. 
- A lot of our components are connected to the redux store. Wrapping the component under test with `Provider` ensures that the `store` is available inside the component during the test.
- A lot of our components receive `history` and/or `location` information from the `connected-react-router` component.  Wrapping the component under test with `ConnectedRouter` ensures that theses are available inside the component during the test.

Not all components require all of these. There are many combinations (e.g., a component might need `i18n` and `store`, but not the router info, whereas another might need `store` and the router info). We could complicate this `render` helper by making it smart enough to only wrap with the necessary providers, but that would make it more complex and would also require the tests to configure that. I am taking the approach here of just always wrapping everything with all 3 HOCs. I *think* this is okay, and shouldn't invalidate any tests we write, but I am interested in feedback from the team about whether you think this is a reasonable way to go, or if we should make sure that whatever component we are testing is *only* wrapped with the HOCs that it actually uses in the app.

**Comment by bobsilverberg** [comment]
I decided to start with this file because it needs both a `store` and a connected router, as well as using the `translate()` HOC, and I wanted to figure out how to accomplish all that without shallow rendering. I created a helper `render` function that can do all those things.

File: tests/unit/amo/components/TestAddonMoreInfo.js [hunk 2]
```diff
@@ -46,42 +43,43 @@ describe(__filename, () => {
...
-    const root = render({ addon: null });
+    const { queryByClassName } = render({ addon: null });
 
     // These fields will be visible during loading since
     // they will always exist for the loaded add-on.
-    expect(root.find('.AddonMoreInfo-last-updated')).toHaveLength(1);
+    expect(screen.getByText('Last updated')).toBeInTheDocument();
 
-    expect(root.find(LoadingText)).toHaveLength(1);
+    expect(queryByClassName('LoadingText')).toHaveLength(1);
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestAddonMoreInfo.js:68]: We have to use `queryByClassName` because there is nothing added to the DOM for `LoadingText` except a `<span>` with a class of `LoadingText`. We have to get access to `queryByClassName` by destructuring `render` because it is a custom query that I created and therefore is not added to the `screen` object. I am looking into whether there is a way to add our custom queries to the `screen` object, but I haven't found a way yet.

File: tests/unit/amo/components/TestAddonMoreInfo.js [hunk 3]
```diff
@@ -46,42 +43,43 @@ describe(__filename, () => {
...
 
     // These fields will not be visible during loading
     // since they may not exist.
-    expect(root.find('.AddonMoreInfo-links')).toHaveLength(0);
-    expect(root.find('.AddonMoreInfo-license')).toHaveLength(0);
-    expect(root.find('.AddonMoreInfo-privacy-policy')).toHaveLength(0);
-    expect(root.find('.AddonMoreInfo-version')).toHaveLength(0);
-    expect(root.find('.AddonMoreInfo-version-history')).toHaveLength(0);
-    expect(root.find('.AddonMoreInfo-eula')).toHaveLength(0);
+    expect(screen.queryByText('Add-on Links')).not.toBeInTheDocument();
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestAddonMoreInfo.js:72]: `getByText` will throw if nothing is found, which is why we need to use `queryByText` if we want to assert that something is not present in the document.

File: tests/unit/amo/components/TestAddonMoreInfo.js [hunk 4]
```diff
@@ -259,117 +255,116 @@ describe(__filename, () => {
...
-    const link = root.find('.AddonMoreInfo-license-link');
-
-    expect(root.find('.AddonMoreInfo-license')).toHaveProp('term', 'License');
-    expect(link.children()).toIncludeText(licenseName);
-    expect(link).toHaveProp('to', `/addon/${addon.slug}/license/`);
+    expect(screen.getByText('License')).toBeInTheDocument();
+    const link = screen.getByText(licenseName);
+    expect(link).toHaveAttribute(
+      'href',
+      `/en-US/android/addon/${addon.slug}/license/`,
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestAddonMoreInfo.js:264]: We need to add `/en-US/android/` to the beginning of the link in this assertion as that's what gets rendered. The previous test was just checking the `to` prop of the `Link` component, so it didn't need to add those to the assertion.

We tend to be very explicit in addons-frontend tests, so if we want to continue with that practice we should probably set up the data in the test which sets a `lang` and `app`, and then use those values in this assertion. I think, though, as we are rewriting these tests, that we don't really need to do that. I think it's fine to just hardcode `/en-US/android/` here because we know that will be the value based on the defaults for the test. I think this is consistent with how a lot of tests are written for addons-server, where, rather than explicitly setting up all the data in the test, the tests rely on existing fixtures and just assert hardcoded values in the test. As this isn't meant to be a test of the `Link` component, I think it's fine to do what I have done here, but I did want to call out that it's not normally what we would do in an addons-frontend test, so we can discuss if we are okay with moving in this direction.

**Comment by bobsilverberg** [comment]
Note that the test failures here are related to a problem with the `is` locale, and that the new test I wrote is in fact passing. 😄

**Comment by bobsilverberg** [comment]
I have flagged all of you for review of this draft PR so we can discuss some issues related to how we want to write these tests in this PR, which will give us some useful context.

File: tests/unit/helpers.js [hunk 5]
```diff
@@ -1470,3 +1474,68 @@ export const createExperimentData = ({
...
+const getByTagName = (container, tagName) => {
+  const elements = queryByTagName(container, tagName);
+  if (!elements.length) {
+    return null;
+  }
+  if (elements.length > 1) {
+    throw new Error(`More than one element found for tag ${tagName}`);
+  }
+  return elements[0];
+};
```
Comment (eviljeff) [inline tests/unit/helpers.js:1506]: Do we need this helper when there is `queryByRole` already?

File: tests/unit/helpers.js [hunk 6]
```diff
@@ -1470,3 +1474,68 @@ export const createExperimentData = ({
...
+
+const queryByClassName = (container, className) => {
+  return container.querySelectorAll(`.${className}`);
+};
+
+const getByClassName = (container, className) => {
+  const elements = queryByClassName(container, className);
+  if (!elements.length) {
+    return null;
+  }
```
Comment (eviljeff) [inline tests/unit/helpers.js:1486]: in the testing library [api](https://testing-library.com/docs/queries/about#types-of-queries), `get` would raise for 0 matches.  - it'd be `queryBy` to get `null` for empty.  I.e. this function should be called `queryByClassName` to be consistent (though that conflicts with the raw function you already have with that name).

File: tests/unit/amo/components/TestAddonMoreInfo.js (hunk 2 above)
Comment (eviljeff) [reply tests/unit/amo/components/TestAddonMoreInfo.js:68]: I guess we could have our own `screen` object that's a wrapper around react testing libraries `screen` object?  (Or was that tried and didn't work)

**Review by eviljeff — APPROVED** [review]
The tests themselves seem fine to me - the changes might be more annoying with a lower level component I suppose.

I just have the comments about the naming and necessity of some of the helper functions.

File: tests/unit/helpers.js [hunk 7]
```diff
@@ -1470,3 +1474,68 @@ export const createExperimentData = ({
...
+  return elements[0];
+};
+
+export const render = (ui, options = {}) => {
+  const internalOptions = {
+    history: addQueryParamsToHistory({ history: createMemoryHistory() }),
+    i18n: fakeI18n(),
+    store: dispatchClientMetadata().store,
+    rtlOptions: {},
```
[the remark this answers is not in the document]
Comment (willdurand) [reply tests/unit/helpers.js:1513]: It would be good to change that indeed

Comment (willdurand) [inline tests/unit/helpers.js:1513]: Why is this needed here?

File: tests/unit/helpers.js [hunk 8]
```diff
@@ -1470,3 +1474,68 @@ export const createExperimentData = ({
...
+  };
+  const rtlOptions = {
+    queries: {
+      ...queries,
+      queryByClassName,
+      queryByTagName,
+      getByClassName,
+      getByTagName,
+    },
+    ...internalOptions.rtlOptions,
```
Comment (willdurand) [inline tests/unit/helpers.js:1524]: This is a bit confusing. `internalOptions` contains `rtlOptions` that can somehow be updated by `options` and then it ends up here?

Maybe define the options explicitly since `internalOptions` does not seem to be passed anywhere as an object:

```js
export const render = (ui, options = {}) => {
  const i18n = options.i18n || fakeI18n();
  const history = options.history || addQueryParamsToHistory({ history: createMemoryHistory() });
  const store = options.store || dispatchClientMetadata().store;
  const rtlOptions = {
    queries: {
      ...queries,
      queryByClassName,
      queryByTagName,
      getByClassName,
      getByTagName,
    },
    ...options.rtlOptions,
  };

  // ...
```

File: tests/unit/helpers.js (hunk 7 above)
Comment (bobsilverberg) [reply tests/unit/helpers.js:1513]: To be honest, I'm not sure. 😉 I think it is a remnant of when I was originally working on this stuff last year. I wanted to make it possible to override the options that eventually get passed to react-testing-library's `render` function, but I'm not sure if we will ever need to do that, and if there is a rare edge case, we could just do everything manually without the helper. I will look into this a bit more, but maybe we can just remove this.

File: tests/unit/helpers.js (hunk 8 above)
Comment (bobsilverberg) [reply tests/unit/helpers.js:1524]: Yes, I think this is better. It also means one can omit just some options and get the default values.

File: tests/unit/helpers.js (hunk 6 above)
Comment (bobsilverberg) [reply tests/unit/helpers.js:1486]: I agree that we should be more consistent with these, and `getBy` should raise for 0 matches. I will update these to make them work that way.

File: tests/unit/helpers.js (hunk 5 above)
Comment (bobsilverberg) [reply tests/unit/helpers.js:1506]: I was using this to locate elements based on tags, such as `<circle>` and `<path>`, which were in my proof-of-concept, but don't appear in this PR. Those are not valid roles, so tag and role are not synonymous.

File: tests/unit/helpers.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply tests/unit/helpers.js:1513]: I am thinking we can get rid of this, but if we cannot, I will change it, probably to `testOptions`.

File: tests/unit/helpers.js (hunk 5 above)
Comment (eviljeff) [reply tests/unit/helpers.js:1506]: 😞  I was hoping `queryByRole` would fall back to tag names, but it wasn't clear from the docs

File: tests/unit/amo/components/TestAddonMoreInfo.js (hunk 2 above)
Comment (bobsilverberg) [reply tests/unit/amo/components/TestAddonMoreInfo.js:68]: I got this to work and it's included in the next commit.

File: tests/unit/helpers.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply tests/unit/helpers.js:1513]: We actually don't need to accept an `options` object, so I have removed this, but I have also changed anywhere else I was using `rtl` to mean `react-testing-library` and replaced it with `library`. It's only currently used in internal variable names and not anywhere as passed arguments.

Comment (bobsilverberg) [reply tests/unit/helpers.js:1513]: We can remove it, and I have done so.
